torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -0,0 +1,406 @@
+  std::unique_lock<std::mutex> lock(m_);
+  cache_[entry->key] = std::move(entry);
+  cacheCV_.notify_all();
+}
+
+void ProcessGroupGloo::createAlgorithm(AlgorithmEntry& entry) {
+  const auto& key = entry.key;
+  switch (key.collectiveType) {
+    case CollectiveType::ALLREDUCE:
+      GENERATE_ALL_TYPES(key.type->scalarType(), createAllreduce, entry);
+      return;
+    case CollectiveType::BROADCAST:
+      GENERATE_ALL_TYPES(key.type->scalarType(), createBroadcast, entry);
+      return;
+  }
+
+  throw std::runtime_error("Unhandled collective type");
+}
+
+template <typename T>
+void ProcessGroupGloo::createAllreduce(AlgorithmEntry& entry) {
+  const auto& key = entry.key;
+
+  // Create algorithm against first context
+  auto& context = contexts_[0];
+  entry.algorithm = std::unique_ptr<::gloo::Algorithm>(
+      new ::gloo::AllreduceHalvingDoubling<T>(
+          context,
+          getDataPointers<T>(entry.src),
+          entry.src[0].numel(),
+          reductionFunction<T>(key.reduceOp)));
+}
+
+template <typename T>
+void ProcessGroupGloo::createBroadcast(AlgorithmEntry& entry) {
+  const auto& key = entry.key;
+
+  // Create algorithm against first context
+  auto& context = contexts_[0];
+  entry.algorithm =
+      std::unique_ptr<::gloo::Algorithm>(new ::gloo::BroadcastOneToAll<T>(
+          context,
+          getDataPointers<T>(entry.src),
+          entry.src[0].numel(),
+          key.srcRank,
+          key.srcTensor));
+}
+
+// Constructs an AlgorithmEntry instance, except for the algorithm
+// itself. It allocates temporary input/output tensors, CUDA streams
+// (if applicable), etcetera. These are lazily allocated and reused
+// for collective calls with the same signature.
+//
+// They cannot be allocated asynchronously because the collective
+// functions need them before queueing their asynchronous work. For
+// example, to work with asynchronous CUDA code, the collective call
+// needs to issue an asynchronous memory copy, and a call to
+// cudaStreamWaitEvent to make it wait for the asynchronous execution
+// of the collective call to complete.
+//
+// Construction of the Gloo algorithm itself it delayed until a thread
+// picks up the work, because it performs I/O and can fail. Any I/O
+// failure must be signaled through the Work future.
+//
+EntryType ProcessGroupGloo::construct(const AlgorithmKey& key) {
+  auto entry = std::unique_ptr<AlgorithmEntry>(new AlgorithmEntry);
+  entry->key = key;
+
+  // Allocate source tensors for this entry
+  auto& srcSizes = key.srcSizes;
+  entry->src.resize(srcSizes.size());
+  for (int i = 0; i < srcSizes.size(); i++) {
+    entry->src[i] = at::zeros(*key.type, at::IntList(srcSizes[i]));
+  }
+
+  return std::move(entry);
+}
+
+EntryType ProcessGroupGloo::checkout(const AlgorithmKey& key) {
+  std::unique_lock<std::mutex> lock(m_);
+
+  // Initialize number of entries for this key.
+  if (cacheCreated_.count(key) == 0) {
+    cacheCreated_.emplace(key, 0);
+  }
```

Comment:
This is unnecessary. Doing `cacheCreated_[key]` is guaranteed to insert a `0` value and return it if the `key` has been missing